### PR TITLE
Dynamic command registry

### PR DESCRIPTION
**Pull request by solnic**
Just like with mapper registry, this turns command registry into a dynamic registry that can compile commands on the fly and encapsulate access to commands in `Relation#command`. All relations have access to their command registries as a consequence.

### CODE REVIEW

File: core/lib/rom/command_compiler.rb [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    #   @param adapter [Symbol] The adapter identifier
+    #   @param ast [Array] The AST representation of a relation
+    #   @param plugins [Array<Symbol>] A list of optional command plugins that should be used
+    #   @param meta [Hash] Meta data for a command
+    #
+    #   @return [Command, CommandProxy]
+    #
+    # @api private
+    def [](*args)
+      fetch_or_store(args.hash) do
```
Comment (flash-gordon) [inline core/lib/rom/command_compiler.rb]: @solnic didn't we lose `freeze` somewhere?

Comment (solnic) [inline core/lib/rom/command_compiler.rb]: Freezing what exactly?

Comment (flash-gordon) [inline core/lib/rom/command_compiler.rb]: the command class :)

Comment (solnic) [inline core/lib/rom/command_compiler.rb]: Oh I didn't know we freeze command classes. IIRC I *planned* to add freezing them (same with mappers and relations).

Comment (flash-gordon) [inline core/lib/rom/command_compiler.rb]: remember this https://github.com/hanami/model/pull/405

Comment (flash-gordon) [inline core/lib/rom/command_compiler.rb]: @solnic I just wanted to know if you still remember it :)

Comment (solnic) [inline core/lib/rom/command_compiler.rb]: Ah OK :) yeah I do remember, I'll do it once I'm done with local caches refactor in #408

**Pull request merged**